Use `changelogs/release.yaml` for the version number source everywhere

anoadragon453 opened the pull request:
Currently both the top banner of the site and the top paragraph on the Changelogs page pull the current version number of the spec from `config.toml`, i.e:

https://github.com/matrix-org/matrix-doc/blob/6c6aa0b3d6eb6edc65c68ad78072a6733998969b/layouts/shortcodes/changelog/changelog-description.html#L8-L15

Note the `{{ .Site.Params.version.number }}` bit.

However, we already have a method of specifying the version number of a release build of the spec: `changelogs/release.yaml`:

https://github.com/matrix-org/matrix-doc/blob/6c6aa0b3d6eb6edc65c68ad78072a6733998969b/layouts/shortcodes/changelog/changelog-changes.html#L1-L16

This commit changes the site banner and Changelogs page to use `tag` in `changelogs/release.yaml` for the version number of the spec, instead of having two different ways to do things. Note that if we're building an unstable version of the spec, then `changelogs/release.yaml` is never read from.

richvdh commented on layouts/partials/navbar.html:
```diff
@@ -45,7 +45,10 @@
     {{ if ne $status "unstable"}}
-        {{ $ret = .Site.Params.version.number }}
+        {{ $path := path.Join "changelogs" }}
+        {{ $release_info := readFile (path.Join $path "release.yaml") | transform.Unmarshal }}
+
+        {{ $ret = $release_info.tag }}
```
what's the difference between `=` and `:=` here?

anoadragon453 replied:
Good question. The syntax is analogous to golang, where `:=` means an initial assignment to a variable, and `=` is a re-assignment. `$ret = $release_info.tag` is only allowed here as `$ret` is already defined.

https://golang.org/doc/effective_go#redeclaration has the details for the operators in go, which I *think* roughly translate to hugo. [Hugo's docs](https://gohugo.io/templates/introduction/#variables) are a little light in this regard.

richvdh replied:
ohhhhhh. I was assuming it was analogous to Makefile syntax. Don't ask. Makes sense now :+1:

richvdh commented on layouts/shortcodes/changelog/changelog-description.html:
```diff
@@ -8,8 +8,14 @@
+
+{{ $path := path.Join "changelogs" }}
+{{ $release_info := readFile (path.Join $path "release.yaml") | transform.Unmarshal }}
+<p>This is version <strong>{{ $release_info.tag }}</strong> of the Matrix specification.</p>
```
I don't know what `slice "version" does in `navbar.html`. Do we also need it here?

anoadragon453 replied:
[slice](https://gohugo.io/functions/slice/) creates an array from the space-separated argument it's given, while [delimit](https://gohugo.io/functions/delimit/) takes an array and a delimiter and spits out a string. This is effectively just a fancy way of saying `$ret = "version " + $ret`. Seems like this is *the way* to concat strings in hugo.

We only need it in the navigation bar as only it shows "version" before the version number:

![image](https://user-images.githubusercontent.com/1342360/128164725-ffb1d6e4-e36e-4ea8-ab5a-08a485176b97.png)

Whereas the nav page just includes the number in a paragraph.

richvdh reviewed (approved):
I'm not really familiar with this templating system, whatever it is - but it looks plausible, as long as you've tested it :)

richvdh commented:
The thought occurs to me: if the `unstable` flag comes from `config.toml`, shouldn't the version number? In other words, should we go the other way and make `changelog-changes` use the value from `config.toml` rather than `release.yaml`?

anoadragon453 commented:
> I'm not really familiar with this templating system, whatever it is - but it looks plausible, as long as you've tested it :)

Oh right - I should've mentioned. It's hugo's own templating system. Looks quite similar to jinja2. https://gohugo.io/templates/introduction/

> The thought occurs to me: if the `unstable` flag comes from `config.toml`, shouldn't the version number? In other words, should we go the other way and make `changelog-changes` use the value from `config.toml` rather than `release.yaml`?

The historical context is in https://github.com/matrix-org/matrix-doc/pull/2992 (note that we switched from individual `release.yaml` files to a single one in https://github.com/matrix-org/matrix-doc/pull/2992 after the global version number proposal). I believe the original reason for keeping it in a file versus `config.toml` was that we originally had to deal with multiple versions of the spec. Nowadays that's not the case.

It would certainly be easier in templates to pull a parameter from the config than parsing yaml files everywhere, so perhaps we should move the date and time into `config.yaml`. Note that the Changelog page only shows the current release version and the changes since the last release. We just export a different release per git tag, so there'll never be a need for multiple `release.yaml` files.

anoadragon453 commented:
@richvdh I've updated things to use the information in `config.toml` everywhere instead of having a separate `changelogs/release.yaml` file. Looks good on a test! :slightly_smiling_face:

richvdh commented on layouts/partials/navbar.html:
```diff
@@ -46,10 +46,9 @@
 
-        {{ $ret = $release_info.tag }}
-        {{ $ret = delimit (slice "version" $ret) " " }}
+		{{/* produces a string similar to "version v1.5" */}}
+		{{ $ret = delimit (slice "version" " v" .Site.Params.version.major "." .Site.Params.version.minor) "" }}
```
indentation is inconsistent?

richvdh replied:
```suggestion
		{{/* produces a string similar to "version v1.5" */}}
		{{ $ret = delimit (slice "version v" .Site.Params.version.major "." .Site.Params.version.minor) "" }}
```

richvdh reviewed (approved):
lgtm otherwise. Presumably this necessitates a change to #3295?

anoadragon453 commented:
> Presumably this necessitates a change to #3295?

Yep! Hope you don't mind if I do so via a rebase...
